juju/juju
# worker/caasoperator: download & unpack charm

## Description of change

When the operator starts up, it checks if it
has a charm directory. If it does not, it will
ask for the application's charm URL and SHA256
hash, download and verify, and then unpack it
into the charm directory. The agent also sets
its status to maintenance during that process,
and active later.

Also:
 - move Dockerfile and requirements.txt under the caas directory
 - change CAAS operator agent to never use lumberjack, always log to stderr (so we can use `kubectl logs`)

## QA steps

1. minikube start
2. juju bootstrap localhost
3. juju add-caas kubernetes k8s
4. juju add-model k8s k8s
5. juju deploy ubuntu
6. eval $(minikube docker-env)
7. make operator-image
8. kubectl set image po/juju-operator-ubuntu juju-operator=juju/caas-jujud-operator:latest
(wait for pod to be recreated)
9. kubectl logs po/juju-operator-ubuntu
(should see via logs that the charm is downloaded and "ready at /var/lib/juju/charm")
10. docker exec -i <container> ls /var/lib/juju/charm
(should see contents of ubuntu charm)

## Documentation changes

None.

## Bug reference

None.
